be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/LaoValidator.scala

```diff
@@ -12,129 +13,241 @@ import ch.epfl.pop.storage.DbActor
+
+          val expectedHash: Hash = Hash.fromStrings(
+            senderPK.toString,
+            stateLao.creation.toString,
+            stateLao.name
+          )
+
+          runChecks(
+            checkTimestampStaleness(
+              rpcMessage,
+              stateLao.creation,
+              validationError(s"stale 'creation' timestamp (${stateLao.creation})")
+            ),
+            checkTimestampOrder(
+              rpcMessage,
+              stateLao.creation,
+              stateLao.last_modified,
+              validationError(s"'last_modified' (${stateLao.last_modified}) timestamp is smaller than 'creation' (${stateLao.creation})")
+            ),
+            checkWitnesses(
+              rpcMessage,
+              stateLao.witnesses,
+              validationError(dupWitnessError)
+            ),
+            checkWitnessesSignatures(
+              rpcMessage,
+              stateLao.modification_signatures,
+              stateLao.modification_id,
+              validationError("witness key-signature pairs are not valid for the given modification_id")
+            ),
+            checkId(
+              rpcMessage,
+              expectedHash,
+              stateLao.id,
+              validationError(unexpectedID + " " + stateLao.id.toString)
+            )
+          )
 
-    rpcMessage.getParamsMessage match {
-      case Some(message: Message) =>
-        val data: UpdateLao = message.decodedData.get.asInstanceOf[UpdateLao]
+        case _ => Right(validationErrorNoMessage(rpcMessage.id))
+      }
+    }
 
-        val channel: Channel = rpcMessage.getParamsChannel
+    def validateGreetLao(rpcMessage: JsonRpcRequest): GraphMessage = {
+      def validationError(reason: String): PipelineError = super.validationError(reason, "GreetLao", rpcMessage.id)
+
+      rpcMessage.getParamsMessage match {
+        case Some(message: Message) =>
+          val (greetLao, _, senderPK, channel) = extractData[GreetLao](rpcMessage)
+          val expectedHash = rpcMessage.extractLaoId
+
+          runChecks(
+            checkId(
+              rpcMessage,
+              expectedHash,
+              greetLao.lao,
+              validationError("unexpected id, was " + greetLao.lao + " but expected " + expectedHash)
+            ),
+            checkMsgSenderKey(
+              rpcMessage,
+              greetLao.frontend,
+              senderPK,
+              validationError(s"unexpected frontend public key ${greetLao.frontend}")
+            ),
+            checkStringPattern(
+              rpcMessage,
+              greetLao.address,
+              addressRegPat,
+              validationError(s"invalid address ${greetLao.address}")
+            ),
+            checkChannel(
+              rpcMessage,
+              channel,
+              Channel(s"${Channel.ROOT_CHANNEL_PREFIX}${greetLao.lao}"),
+              validationError(s"trying to write an GreetLao message on wrong channel $channel")
```

Comment:
`"trying to write **a** GreetLao message on wrong channel $channel"` (sorry it was my error ..)